tooling/SecretManager.VisualStudio/ProjectLocalSecretsManager.cs

```diff
@@ -0,0 +1,107 @@
+
+        private readonly IProjectProperties _projectProperties;
+
+        public ProjectLocalSecretsManager(IProjectProperties properties)
+        {
+            _projectProperties = properties;
+        }
+
+        public string SanitizeName(string name) => name;
+
+        public IReadOnlyCollection<char> GetInvalidCharactersFrom(string name) => Array.Empty<char>();
+
+        public Task AddSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+            => SetSecretAsync(name, value, cancellationToken);
+
+        public async Task SetSecretAsync(string name, string value, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+
+            var store = await GetOrCreateStoreAsync();
+
+            store.Set(name, value);
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                store.Save();
+            }
+        }
+
+        public async Task<string> GetSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+            var store = await GetOrCreateStoreAsync();
+            return store[name];
+        }
+
+        public async Task<IReadOnlyCollection<string>> GetSecretNamesAsync(CancellationToken cancellationToken = default)
+        {
+            var store = await GetOrCreateStoreAsync();
+            return store.ReadOnlyKeys;
+        }
+
+        public async Task<IReadOnlyDictionary<string, string>> GetSecretsAsync(CancellationToken cancellationToken = default)
+        {
+            return await GetOrCreateStoreAsync();
+        }
+
+        public async Task<bool> RemoveSecretAsync(string name, CancellationToken cancellationToken = default)
+        {
+            EnsureKeyNameIsValue(name);
+
+            var store = await GetOrCreateStoreAsync();
+
+            var result = store.Remove(name);
+
+            if (!cancellationToken.IsCancellationRequested)
+            {
+                store.Save();
+            }
+
+            return result;
+        }
+
+        private void EnsureKeyNameIsValue(string name)
+        {
+            if (name == null)
+            {
+                throw new ArgumentNullException(nameof(name));
+            }
+
+            if (string.IsNullOrEmpty(name))
+            {
+                throw new ArgumentException(nameof(name));
+            }
+        }
+
+        private async Task<SecretsStore> GetOrCreateStoreAsync()
+        {
+            var userSecretsId = await _projectProperties.GetEvaluatedPropertyValueAsync(UserSecretsPropertyName);
+
+            if (string.IsNullOrEmpty(userSecretsId))
+            {
+                userSecretsId = Guid.NewGuid().ToString();
+                await _projectProperties.SetPropertyValueAsync(UserSecretsPropertyName, userSecretsId);
+            }
+
+            return new SecretsStore(userSecretsId);
```

Comment:
Can you add `await TaskScheduler.Default;` here so that we know the disk I/O won't be on the UI thread?
Or, if you can't reference MS.VS.Threading.dll, instead of awaiting the task scheduler, you can use this syntax:

```csharp
return await Task.Run(() => new SecretsStore(userSecretsId));
```

Although IMO constructors should be fairly lightweight, so I would personally refactor this to have a `LoadAsync` static factory method on the `SecretsStore`, or keep all the load/save functionality in this outer class and just use a straight-up Dictionary.

Reply:
Similar change for saving the file.

Reply:
Trying to add `await TaskScheduler.Default`. Compielr fails with "Cannot find the interop type that matches the embedded interop type 'Microsoft.VisualStudio.Shell.Interop.IVsTask'. Are you missing an assembly reference?"

Microsoft.VisualStudio.Threading is referenced. Any idea what I'm missing?

Reply:
Sounds like you have a using for `Microsoft.VisualStudio.Shell`. There's a common name conflict there.